### ISSUE TITLE
Property does not exist on type '(Anonymous class)'

### ISSUE THREAD

**Issue by nickelya** [Microsoft/TypeScript#14860]
**TypeScript Version:**
Version 2.2.1
Version 2.3.0-dev.20170325

**Code**

```ts
while (0);

new class {
    methodA() {
        this;
    }
};

new class {
    methodB() {
        this.methodB;
    }
};
```

**Expected behavior:**
No errors.

**Actual behavior:**
testbug.ts(11,14): error TS2339: Property 'methodB' does not exist on type '(Anonymous class)'.

And if you replace `this.methodB` with `this.methodA` it typechecks fine when it should not.
This didn't happen in earlier versions, Version 2.0.0 works as expected.

**Comment by sandersn** [comment]
Caching in control flow analysis goes wrong. It requires a loop and a reference to `this` inside of `methodA`, so it's a fairly specific set of conditions that triggers it.

**Comment by sandersn** [comment]
Fix is up at #15032. The problem was that control flow was treating `this` like a normal variable reference and walking outside the class expression to try to narrow it. Because of caching, the *first* type that control flow analysis found for this (`{ methodA() { }`) was used for `this` everywhere.
